File: server-ui/test/interop/interop_resources.dart
import 'dart:convert';
import 'dart:io';

class InteropResources {
  static String load(String domain, String resource) {
    final file = File('../shared/src/test/resources/interop/$domain/$resource.json');
    if (!file.existsSync()) {
      throw StateError('Interop resource not found: [${file.path}]');
    }
    return file.readAsStringSync();
  }

  static Object? tree(String raw) => _normalize(jsonDecode(raw));

  static Object? _normalize(Object? value) {
    if (value is Map) {
      final out = <String, Object?>{};
      for (final entry in value.entries) {
        final key = entry.key as String;
        if (_ignoredKeys.contains(key) || entry.value == null) continue;
        out[key] = _normalize(entry.value);
      }
      return out;
    }
    if (value is List) {
      return value.map(_normalize).toList();
    }
    if (value is String) {
      if (_uuidPattern.hasMatch(value)) return value.toLowerCase();
      if (_dateTimePattern.hasMatch(value)) {
        final parsed = DateTime.tryParse(value);
        if (parsed != null) return parsed.toUtc().toIso8601String();
      }
      return value;
    }
    return value;
  }

  static const Set<String> _ignoredKeys = {
    'next_invocation',
    'use_query_string',
    'context',
    'additional_config',
  };

  static final RegExp _uuidPattern = RegExp(
    r'^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$',
  );

  static final RegExp _dateTimePattern = RegExp(
    r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:?\d{2})?$',
  );
}

File: client-ios/lib/Tests/StasisClientLibTests/Interop/InteropResources.swift
import Foundation

enum InteropResources {
    static func load(domain: String, resource: String) -> Data {
        let url = baseDirectory
            .appendingPathComponent(domain)
            .appendingPathComponent("\(resource).json")
        do {
            return try Data(contentsOf: url)
        } catch {
            fatalError("Interop resource not found: [\(url.path)]")
        }
    }

    static func tree(_ data: Data) -> NSDictionary {
        do {
            guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                fatalError("Interop resource is not a JSON object")
            }
            guard let normalized = normalize(dict) as? NSDictionary else {
                fatalError("Failed to normalize interop resource as a dictionary")
            }
            return normalized
        } catch {
            fatalError("Failed to parse interop resource: \(error)")
        }
    }

    private static func normalize(_ value: Any) -> Any {
        switch value {
        case let dict as [String: Any]:
            return dict.reduce(into: [String: Any]()) { result, pair in
                guard !ignoredKeys.contains(pair.key), !(pair.value is NSNull) else { return }
                result[pair.key] = normalize(pair.value)
            }
        case let array as [Any]:
            return array.map(normalize)
        case let string as String:
            return uuidPattern.firstMatch(in: string, range: NSRange(string.startIndex..., in: string)) != nil
                ? string.lowercased()
                : string
        default:
            return value
        }
    }

    private static let ignoredKeys: Set<String> = [
        "next_invocation",
        "use_query_string",
        "context",
        "additional_config"
    ]

    private static let uuidPattern: NSRegularExpression = {
        do {
            return try NSRegularExpression(
                pattern: "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
            )
        } catch {
            fatalError("Invalid UUID regex: \(error)")
        }
    }()

    private static let baseDirectory: URL = {
        URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("Resources")
            .appendingPathComponent("interop")
    }()
}
